### PR TITLE
trans: Exit earlier from base::trans_crate() when compiling rmeta crates.

### PR DESCRIPTION
**Pull request by michaelwoerister**
Fixes https://github.com/rust-lang/rust/issues/38964.
r? @eddyb 
cc @nrc

### CODE REVIEW

File: src/librustc_trans/back/write.rs [hunk 1]
```diff
@@ -667,7 +667,8 @@ pub fn run_passes(sess: &Session,
 
     // Sanity check
     assert!(trans.modules.len() == sess.opts.cg.codegen_units ||
-            sess.opts.debugging_opts.incremental.is_some());
+            sess.opts.debugging_opts.incremental.is_some() ||
+            output_types.contains_key(&OutputType::Metadata));
```
Comment (KalitaAlexey) [inline src/librustc_trans/back/write.rs]: Are you sure that there is no chance that `output_types` may contain not only `Metadata`?

File: src/librustc_trans/base.rs [hunk 2]
```diff
@@ -1145,6 +1145,23 @@ pub fn trans_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     };
     let no_builtins = attr::contains_name(&krate.attrs, "no_builtins");
 
+    // Skip crate items and just output metadata in -Z no-trans mode.
+    if tcx.sess.opts.debugging_opts.no_trans ||
+       tcx.sess.opts.output_types.contains_key(&config::OutputType::Metadata) {
```
Comment (eddyb) [inline src/librustc_trans/base.rs]: I'd rather reverse this sort of condition and instead have `!output_types.iter().any(|o| o.needs_codegen())`.

File: src/librustc_trans/back/write.rs (hunk 1 above)
Comment (michaelwoerister) [inline src/librustc_trans/back/write.rs]: This is just replicating the condition from here: 
https://github.com/rust-lang/rust/blob/47965f51e6d91ae33bf2b0da0f5510f03766feac/src/librustc_trans/base.rs#L1186

It's not an important assertion anyway.

**Comment by eddyb** [comment]
In general, it'd be neat if you could generate an `.o` and an `.rmeta` and package your own `.rlib`.
Or `cargo` and/or incremental recompilation could do that to avoid the packaging on intermediary libs.
Linking might be a bit silly but you should be able to do it... somehow.

File: src/librustc_trans/base.rs (hunk 2 above)
Comment (michaelwoerister) [inline src/librustc_trans/base.rs]: I just so that there's already `OutputTypes::should_trans()` which does what's needed here. I'll change it.

Comment (michaelwoerister) [inline src/librustc_trans/base.rs]: Or maybe not. What was that with `metadata,bin` being special?

Comment (michaelwoerister) [inline src/librustc_trans/base.rs]: No, it looks like that should work.

**Comment by michaelwoerister** [comment]
Updated. Both comments should be addressed now.

File: src/librustc_trans/base.rs [hunk 3]
```diff
@@ -1145,6 +1145,23 @@ pub fn trans_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     };
     let no_builtins = attr::contains_name(&krate.attrs, "no_builtins");
 
+    // Skip crate items and just output metadata in -Z no-trans mode.
+    if tcx.sess.opts.debugging_opts.no_trans ||
+       !tcx.sess.opts.output_types.should_trans() {
```
Comment (eddyb) [inline src/librustc_trans/base.rs]: I wonder if we should even keep `-Zno-trans`.

**Comment by alexcrichton** [comment]
@bors: r-

I believe this causes `-Zno-trans` to ICE. I've [fixed this](https://github.com/rust-lang/rust/pull/39199/commits/3fbe864375c9f2e9ba07b2891c3118b9afc134ec#diff-a3b24dbe2ea7c1981f9ac79f9745f40a) in the [rollup](https://github.com/rust-lang/rust/pull/39199) which I hope to land today, though, so it may not be necessary to update this PR.

**Pull request merged**